### PR TITLE
refactor(feature-flags): add intersection tests; structure refinement

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue number:** #3531

## Summary

Adds remaining changes before releasing new intersection rule actions `ANY_IN_VALUE`, `ALL_IN_VALUE`, `NONE_IN_VALUE`.

### Changes

> Please provide a summary of what's being changed
* [x] Test new operators
  * [x] `ANY_IN_VALUE`
  * [x] `ALL_IN_VALUE`
  * [x] `NONE_IN_VALUE`
* [x] Use `context_value` and `condition_value` fn arguments
* [x] Static type function arguments
* [x] Docstring functions
* [x] Split context vs schema value validation
  * [x] Move condition value validation under Schema Validator
  * [x] Raise `ValueError` for non-list context value
* [x] Test invalid schema values
  * [x] `ANY_IN_VALUE`
  * [x] `ALL_IN_VALUE`
  * [x] `NONE_IN_VALUE`
* [x] Add Exception Handler

**Unrelated**

* [x] Load rule action mapping once (we're defining `lambda` actions on every evaluation call)
* [x] Refactor if/elif/elif with a simple dispatcher (`getattr`)
  * [x] Validator noop
  * [x] Validator getattr
  * [x] Validator days of the week
  * [x] Validator modulo range
  * [x] Validator time value
  * [x] Validator datetime value
  * [x] Get rid of redundant validations in time/datetime value
  * [x] Decouple exact exception message in test but specifics
  * [x] Condition keys
* [x] Add docstrings and return to dispatchers
* [x] Add docstring on end-to-end flow

### User experience

> Please share what the user experience looks like before and after this change

**ALL_IN_VALUE**

> `context=["Gerald"]`

```json
{
    "my_feature": {
        "default": false,
        "rules": {
            "tenant_id is in allowed list": {
                "when_match": true,
                "conditions": [
                    {
                        "action": "ALL_IN_VALUE",
                        "key": "tenant_id",
                        "value": [
                            "Łukasz",
                            "Gerald",
                            "Leandro",
                            "Heitor"
                        ]
                    }
                ]
            }
        }
    }
}
```

**ANY_IN_VALUE**

> `context=["Gerald"]`

```json
{
    "my_feature": {
        "default": false,
        "rules": {
            "tenant_id is in allowed list": {
                "when_match": true,
                "conditions": [
                    {
                        "action": "ANY_IN_VALUE",
                        "key": "tenant_id",
                        "value": [
                            "Łukasz",
                            "Gerald",
                            "Leandro",
                            "Heitor"
                        ]
                    }
                ]
            }
        }
    }
}
```

**NONE_IN_VALUE**

> `context=["Simon"]`

```json
{
    "my_feature": {
        "default": false,
        "rules": {
            "tenant_id is in allowed list": {
                "when_match": true,
                "conditions": [
                    {
                        "action": "NONE_IN_VALUE",
                        "key": "tenant_id",
                        "value": [
                            "Łukasz",
                            "Gerald",
                            "Leandro",
                            "Heitor"
                        ]
                    }
                ]
            }
        }
    }
}
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [ ] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [ ] I have performed a self-review of this change
* [ ] Changes have been tested
* [ ] Changes are documented
* [ ] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by heitorlessa** [comment]
Leaving these for another two PRs

* [ ] Perf improvement
	* [ ] Validate same schema once
	* [ ] Identify perf bottleneck w/ feature flags (slowest test in our functional suite)
* [ ] Revamp docs
	* [ ] Revamp user guide (add schema examples, add diagrams, remove excessive banners, etc.)
	* [ ] Revamp Schema Validation errors (provide actionable examples)

**Review by heitorlessa — COMMENTED** [review]
had a peer review with @rubenfonseca and will make the following changes before: 

- Add debug on `noop` method
- Use `cachedproperty` for `TimeValues`

**Comment by heitorlessa** [comment]
Leaving to a separate PR

* [ ] Validate schema once (hash)
* [ ] Revamp docs (add validation_exception_handler, add schema examples, add diagram, remove excessive banners etc)

File: aws_lambda_powertools/utilities/feature_flags/feature_flags.py [hunk 1]
```diff
@@ -350,7 +391,29 @@ def get_enabled_features(self, *, context: Optional[Dict[str, Any]] = None) -> L
 
         return features_enabled
 
-    def exception_handler(self, exc_class: Exception | list[Exception]):
+    def validation_exception_handler(self, exc_class: Exception | list[Exception]):
+        """Registers function to handle unexpected exceptions when evaluating flags.
```
Comment (rubenfonseca) [inline aws_lambda_powertools/utilities/feature_flags/feature_flags.py]:
```suggestion
        """Registers function to handle unexpected validation exceptions when evaluating flags.
```

**Review by rubenfonseca — DISMISSED** [review]
Just left a NP

**Review by rubenfonseca — APPROVED** [review]

**Pull request merged**